src/node/interfaces.cpp

```diff
@@ -860,6 +860,25 @@ class MinerImpl : public Mining
         return tip->GetBlockHash();
     }
 
+    std::pair<uint256, int> waitTipChanged(MillisecondsDouble timeout) override
+    {
+        uint256 previous_hash{WITH_LOCK(::cs_main, return chainman().ActiveChain().Tip()->GetBlockHash();)};
+
+        auto deadline = std::chrono::steady_clock::now() + timeout;
+        {
+            WAIT_LOCK(g_best_block_mutex, lock);
+            while (!chainman().m_interrupt && std::chrono::steady_clock::now() < deadline) {
+                auto check_time = std::chrono::steady_clock::now() + std::min(timeout, MillisecondsDouble(1000));
```

Comment:
I changed this in the last push, previously it would always wait for one second each "round" even if the deadline was shorter.